app/models/concerns/attribute_blockable.rb

```diff
@@ -0,0 +1,379 @@
+          # For example: form_email maps to :email, form_email_domain maps to :email_domain
+          model_class = @records.first.class
+          blockable_config = model_class.blockable_attributes.find { |attr| attr[:blockable_method] == method_name.to_sym }
+          attribute_type = blockable_config ? blockable_config[:object_type] : method_name.to_sym
+
+          scope = BLOCKED_OBJECT_TYPES.fetch(attribute_type, :all)
+          blocked_objects_by_value = BlockedObject.send(scope).find_objects(values).index_by(&:object_value)
+
+          @records.each do |record|
+            value = record.send(method_name)
+            blocked_object = blocked_objects_by_value[value]
+            record.blocked_by_attributes[method_name] = blocked_object
+          end
+        end
+      end
+  end
+
+  module ClassMethods
+    # Defines blockable attribute methods for the given attribute.
+    #
+    # Generates the following instance methods:
+    # - +blocked_by_{method}?+ - Returns true if the value is currently blocked
+    # - +blocked_by_{method}_at?+ - Returns true if blocked_at timestamp is present
+    # - +blocked_by_{method}_at+ - Returns the blocked_at timestamp or nil
+    # - +block_by_{method}_created_at+ - Returns when the block was created
+    # - +block_by_{method}_expires_at+ - Returns when the block expires, or nil
+    # - +blocked_{pluralized_method}_objects+ - Returns BlockedObject records
+    # - +blocked_{pluralized_method}+ - Returns array of blocked values
+    # - +block_by_{method}!+ - Blocks the attribute value
+    # - +unblock_by_{method}!+ - Unblocks the attribute value
+    # - +blocked_by_{method}_object+ - Returns the BlockedObject record
+    #
+    # @param blockable_method [Symbol, String] The method name to make blockable
+    # @param object_type [Symbol, String, nil] The BlockedObject type to use (defaults to blockable_method)
+    #
+    # @example Basic usage
+    #   attr_blockable :email
+    #   # user.blocked_by_email?
+    #   # user.blocked_by_email_at
+    #   # user.block_by_email!(by_user_id: current_user.id)
+    #
+    # @example With custom attribute mapping
+    #   attr_blockable :form_email, object_type: :email
+    #   # Uses 'email' BlockedObject type but creates form_email methods
+    #
+    # @example Blocking with expiration
+    #   user.block_by_email!(
+    #     by_user_id: admin.id,
+    #     expires_in: 30.days
+    #   )
+    def attr_blockable(blockable_method, object_type: nil)
+      object_type ||= blockable_method
+
+      define_method("blocked_by_#{blockable_method}_at?") { (blocked_object_by_method(object_type, blockable_method:)&.blocked_at).present? }
+      define_method("blocked_by_#{blockable_method}?") { blocked_object_by_method(object_type, blockable_method:)&.blocked? || false }
+      define_method("blocked_by_#{blockable_method}_at") { blocked_object_by_method(object_type, blockable_method:)&.blocked_at }
```

Comment:
why do we have both `blocked_by_email_at?` and `blocked_by_email_at` ? would it be possible to retain only the most essential (& used) methods so that we can simplify this module? That will also help understand the module better, currently it's moderately complex.